[tests] Extended controller app tests part one

asmodehn opened the pull request:
## Checklist

- [X] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [X] I have manually tested the changes proposed in this pull request.
- [X] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

## Reference to Existing Issue
This is a part of #303 which was too ambitious. Making a sample_config extend app is trickier than expected, and will be done later.

The goal here is to add a sample_connection extended app, to make sure the case where connection is an extended app is tested.
This happens when someone extends the connection app (to change the connection logic or add complex device command code for instance) and wants to work with it while also using the firmware_upgrader module, which depends on it.

## Description of Changes

Added sample_connection extended app.
Based on #303 changes.
Aiming to get all tests to pass with this as a first step.

To achieve this, we need to refactor some mocks in tests to use a contextmanager for the `*TestCase._mock_connect` attribute instead of a decorator (parsed at parent-class definition time)

Commit:
[change] Add controller config and connection apps

Commit:
[change] Adjust tests and sample apps

Commit:
[change] Remove sample_config for now

nemesifier committed:
Merge branch 'master' into extended_controller_app_tests_part_one

Commit:
[change] Formatting

nemesifier reviewed (commented):
Thanks @asmodehn for your patience in working on this, I scanned the PR and noticed a few small things.

Is the change to `tests/media/floorplan.jpg` intentional? I suggest reverting that.

See also my comments below.

PS: can we close https://github.com/openwisp/openwisp-firmware-upgrader/pull/303 then?

Commit:
[change] Remove comments and commented accounts/ in tests/openwisp/urls.py

asmodehn commented on tests/openwisp2/urls.py:
```diff
@@ -1,18 +1,64 @@
 from django.views.generic import RedirectView
 
+# from openwisp_controller.config.api.urls import get_api_urls as get_config_api_urls
+# from openwisp_controller.config.utils import get_controller_urls
```
Actually, it was practical to have here, but I have this information in the other PR, so I removed it from here.

Commit:
[change] Remove comments regarding config from tests settings

nemesifier commented on tests/openwisp2/sample_connection/tests.py:
```diff
@@ -0,0 +1,122 @@
+class TestNotificationTransaction(BaseTestNotificationTransaction):
+    app_label = "sample_connection"
+
+    def _generic_notification_test(
```
This looks identical to the test above, are you sure it's really needed? We use the TransactionTestCase when DB transactions need to be enabled for the feature to work, I think you can remove this test because it won't be doing anything more than the other test method above.

asmodehn replied:
This is probably cleaner : https://github.com/openwisp/openwisp-controller/pull/1192/files
and should simplify things here... Let me know hat you think

Commit:
[change] Simplify sample_connection.tests.TestNotifications

Commit:
[change] Merge branch 'master' into extended_controller_app_tests_part_one

# Conflicts:
#	openwisp_firmware_upgrader/tests/test_admin.py
#	openwisp_firmware_upgrader/tests/test_selenium.py

Commit:
[change] Remove unused connection_app_label in test_admin.BaseTestAdmin

Commit:
[change] Merge branch 'master' into extended_controller_app_tests_part_one

asmodehn committed:
[change] Merge branch 'master' into extended_controller_app_tests_part_one

asmodehn commented on tests/openwisp2/sample_connection/tests.py:
```diff
@@ -0,0 +1,92 @@
+    config_app_label = "config"
+
+
+class TestConnectionApi(BaseTestConnectionApi):
+    def test_post_deviceconnection_list(self):
+        d1 = self._create_device()
+        self._create_config(device=d1)
+        path = reverse("connection_api:deviceconnection_list", args=(d1.pk,))
+        data = {
+            "credentials": self._get_credentials().pk,
+            "update_strategy": conn_settings.UPDATE_STRATEGIES[0][0],
+            "enabled": True,
+            "failure_reason": "",
+        }
+        with self.assertNumQueries(13):
+            response = self.client.post(path, data, content_type="application/json")
+        self.assertEqual(response.status_code, 201)
```
The parent class does not have that field. it is therefore nor required nor useful to have here.

asmodehn commented on tests/openwisp2/urls.py:
```diff
@@ -1,18 +1,46 @@
+)
 from openwisp_users.api.urls import get_api_urls
 
+from .sample_connection.api import views as connection_api_views
+
 redirect_view = RedirectView.as_view(url=reverse_lazy("admin:index"))
 
-urlpatterns = [
+urlpatterns = []
+
+if os.environ.get("SAMPLE_APP", False):
+    urlpatterns += [
+        path(
+            "",
+            include(("openwisp_controller.config.urls", "config"), namespace="config"),
+        ),
+        path(
+            "api/v1/",
+            include(
+                (
+                    get_connection_api_urls(connection_api_views),
+                    "connection_api",
+                ),
+                namespace="connection_api",
+            ),
+        ),
+    ]
+
```
This is the way we test for SAMPLE_APP in other modules, so lets keep it like that and lets not complicate this more than required.

asmodehn committed:
[change] Update sample_connection app from controller

asmodehn committed:
[change] Reactivate tests fixes for sample_connection

asmodehn committed:
[change] Fix recent test_add_credentials_with_cancelled_upgrade_operation

nemesifier commented on tests/openwisp2/sample_connection/api/views.py:
```diff
@@ -0,0 +1,50 @@
+    pass
+
+
+class DeviceConnenctionListCreateView(BaseDeviceConnenctionListCreateView):
```
Typo here and on the import alias (line 17) and the as_view calls (lines 49-50). Connenction -> Connection (double n). Also deviceconnection_details_view on line 50 should be detail_view.

asmodehn replied:
>Also deviceconnection_details_view on line 50 should be detail_view.

I m duplicating this line : https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/connection/api/views.py#L153 so we ll need to fix the typo in multiple places...

asmodehn replied:
Actually the other one is there too : https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/connection/api/views.py#L133

nemesifier replied:
lol! I'll see what I can do

nemesifier commented on tests/openwisp2/sample_connection/pytest.py:
```diff
@@ -0,0 +1,10 @@
+from openwisp_controller.connection.tests.pytest import (
```
It looks like this file is not wired into any test runner.

asmodehn replied:
Ah good catch. I ll need to review this and add pytest in the test runner

nemesifier reviewed (changes requested):
I scanned the PR and noticed a few things.

The description explains what the code does but not why. What is the real-world use case that requires extending the connection app? Please add that context so it can go into the commit message once merged.

`openwisp_firmware_upgrader/tests/test_admin.py:467` -- The `*args` parameter in `test_add_credentials_with_cancelled_upgrade_operation` is leftover from the old `@mock.patch` decorator pattern. The method now uses context managers via `self._mock_connect` / `self._mock_upgrade`, so `*args` receives nothing. Remove it.

`tests/openwisp2/settings.py` -- Removing the TEMPLATES DIRS entry is unrelated to the connection swap. Was this causing a problem? If not, please revert it to keep the PR focused.

On the mock refactoring -- converting `test_add_credentials_with_cancelled_upgrade_operation` to use `self._mock_connect` / `self._mock_upgrade` is correct. But `TestAdminTransaction` still uses `@mock.patch(_mock_connect, ...)` decorators, which capture the value at parent-class definition time. Subclassing `_mock_connect` will not affect them. This is not a regression but should be acknowledged in the description so reviewers know the scope.

asmodehn committed:
[change] Cleanup redundant *args in test params

asmodehn committed:
[change] Add again TEMPLATES DIRS settings, not related to current changes

asmodehn committed:
[change] Fix typo in sample_connection.api.views

asmodehn committed:
[change] Add pytest to runtests.py for sample_connection tests

asmodehn committed:
[change] Remove redundant (config_)app_label in sample app tests

asmodehn committed:
Merge branch 'master' into extended_controller_app_tests_part_one

asmodehn committed:
[chores] Fix typos after controller update

nemesifier commented on tests/openwisp2/urls.py:
```diff
@@ -1,15 +1,42 @@
+    urlpatterns += [
+        path(
+            "",
+            include(("openwisp_controller.config.urls", "config"), namespace="config"),
```
isn't this redundant with what we have below?

nemesifier commented on tests/openwisp2/urls.py:
```diff
@@ -1,15 +1,42 @@
+urlpatterns += [
     path("admin/", admin.site.urls),
     path("", redirect_view, name="index"),
     path("", include("openwisp_controller.urls")),
```
Here, doesn't this already include the controller URLs? See https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/urls.py

nemesifier commented on tests/openwisp2/urls.py:
```diff
@@ -1,15 +1,42 @@
+            "api/v1/",
+            include(
+                (
+                    get_connection_api_urls(connection_api_views),
```
I understand you're doing this to supply your own views, which seems correct, but wouldn't this be overridden by `path("", include("openwisp_controller.urls")),` below?

nemesifier reviewed (commented):
Looks good, I have a few doubts touching `urls.py`, I am not 100% sure but worth double checking (I aslso double checking on my side).

nemesifier committed:
[chores] Run pytest only in sample-app tests

nemesifier committed:
[chores] Handled urls fixes

nemesifier committed:
[chores] Minor improvements

nemesifier committed:
[chores] Removed debugging traces

nemesifier reviewed (approved):
@asmodehn I pushed some minor fixes, please double check on your end when you can.
